probe/appclient/app_client.go

```diff
@@ -245,6 +247,12 @@ func (c *appClient) publish(r io.Reader) error {
 	req.Header.Set("Content-Encoding", "gzip")
 	req.Header.Set("Content-Type", "application/msgpack")
 	// req.Header.Set("Content-Type", "application/binary") // TODO: we should use http.DetectContentType(..) on the gob'ed
+
+	// Make sure this request is cancelled when we stop the client
+	c.mtx.Lock()
+	req.Cancel = c.quit
+	c.mtx.Unlock()
```

Comment:
I don't think you need the lock (you can read something from another thread, as long as its not modified by any other threads, which this is not).